src/core/services/slot-refresher.ts

```diff
@@ -24,35 +31,123 @@ async function isUAP(): Promise<boolean> {
+	const bidGroup = context.get(`slots.${slotName}.bidGroup`);
+
+	if (!bidGroup) {
+		context.set(`slots.${slotName}.bidGroup`, slotName);
+	}
+
+	communicationService.emit(eventsRepository.BIDDERS_CALL_PER_GROUP, {
+		group: slotName,
+		callback: callback,
+	});
+}
+
+function refreshWhenBackInViewport(adSlot: AdSlot) {
 	function refresh(event) {
 		if (event.slot.getSlotElementId() === adSlot.getSlotName()) {
 			logger(logGroup, `${adSlot.getSlotName()} back in the viewport, refreshing.`, event);
-			GptProvider.refreshSlot(adSlot);
+
+			callBidders(adSlot.getSlotName(), () => {
+				GptProvider.refreshSlot(adSlot);
+			});
+
 			window.googletag.pubads().removeEventListener('slotVisibilityChanged', refresh);
 		}
 	}
 	window.googletag.pubads().addEventListener('slotVisibilityChanged', refresh);
 }
 
-class SlotRefresher {
+export class SlotRefresher {
 	config: Config;
 	slotsInTheViewport: Array<string> = [];
 
 	log(...logValues) {
 		logger(logGroup, ...logValues);
 	}
 
-	refreshSlot(adSlot: AdSlot) {
-		if (!this.config.slots.includes(adSlot.getSlotName())) return;
+	isSlotRefreshable(slotName: string) {
+		return this.config.slots.includes(slotName);
+	}
+
+	/**
+	 * Removes adSlot sizes taller than the first rendered adSlot with the same slotName.
+	 * @param {AdSlot} adSlot - The ad slot to adjust.
+	 */
+	removeHigherSlotSizes(adSlot: AdSlot) {
+		const slotName = adSlot.getSlotName();
+		const slotHeight = adSlot.getElement().clientHeight;
+
+		const slotRefresherConfig = context.get('slotConfig.slotRefresher.sizes');
+		if (!slotRefresherConfig) return;
+
+		const slotHeightLimit = slotRefresherConfig[slotName][1];
+		if (!slotHeightLimit) return;
+
+		const sizeKey = context.get(`slots.${adSlot.getSlotName()}.sizes`);
+
+		if (sizeKey) {
+			context.set(
+				`slots.${adSlot.getSlotName()}.sizes`,
+				sizeKey[0].sizes.filter((size) => size[1] <= slotHeight && size[1] > 3),
+			);
+		}
+		const defaultSizeKey = context.get(`slots.${adSlot.getSlotName()}.defaultSizes`);
+
+		if (defaultSizeKey) {
+			context.set(
+				`slots.${adSlot.getSlotName()}.defaultSizes`,
+				defaultSizeKey.filter((size) => size[1] <= slotHeight && size[1] > 3),
+			);
+		}
+
+		if (adSlot.config.sizes) {
+			adSlot.config.sizes = adSlot.config.sizes.filter((size) => size[1] <= slotHeightLimit);
+		}
 
+		if (adSlot.config.defaultSizes) {
+			adSlot.config.defaultSizes = adSlot.config.defaultSizes.filter(
+				(size) => size[1] <= slotHeightLimit,
+			);
+		}
+
+		const filterCallback = (size: [number, number]) => size[1] <= slotHeightLimit;
+
+		communicationService.emit(eventsRepository.SLOT_REFRESHER_SET_MAXIMUM_SLOT_HEIGHT, {
+			adSlot,
+			filterCallback,
+		});
+	}
+
+	addSlotSizeToContext(adSlot: AdSlot) {
+		const slotWidth = adSlot.getElement().clientWidth;
+		const slotHeight = adSlot.getElement().clientHeight;
+
+		context.set(`slotConfig.slotRefresher.sizes.${adSlot.getSlotName()}`, [slotWidth, slotHeight]);
```

Comment:
as we do not do anything with this value, except pass them, maybe it can be shorten to:
```suggestion
		context.set(`slotConfig.slotRefresher.sizes.${adSlot.getSlotName()}`, [adSlot.getElement().clientWidth, adSlot.getElement().clientHeight]);
```